palace.pyx

```diff
@@ -1158,6 +1162,125 @@ cdef class SourceGroup:
+    """
+    cdef alure.AuxiliaryEffectSlot impl
+
+    def __init__(self, context: Context) -> None:
+        self.impl = (<Context> context).impl.create_auxiliary_effect_slot()
+
+    def __enter__(self) -> AuxiliaryEffectSlot:
+
+        If it is `None`, the object is left uninitialized.
+    """
+    cdef alure.AuxiliaryEffectSlot impl
+
+    def __init__(self, context: Optional[Context]) -> None:
+        if context is None: return
+        self.impl = (<Context> context).impl.create_auxiliary_effect_slot()
+
+    def __enter__(self):
+        return self
+
+    def __exit__(self, exc_type: Optional[Type[BaseException]],
+                 exc_val: Optional[BaseException],
+                 exc_tb: Optional[TracebackType]) -> Optional[bool]:
+        self.destroy()
+
+    def __lt__(self, other: Any) -> bool:
+        if not isinstance(other, AuxiliaryEffectSlot):
+            return NotImplemented
+        return self.impl < (<AuxiliaryEffectSlot> other).impl
+
+    def __le__(self, other: Any) -> bool:
+        if not isinstance(other, AuxiliaryEffectSlot):
+            return NotImplemented
+        return self.impl <= (<AuxiliaryEffectSlot> other).impl
+
+    def __eq__(self, other: Any) -> bool:
+        if not isinstance(other, AuxiliaryEffectSlot):
+            return NotImplemented
+        return self.impl == (<AuxiliaryEffectSlot> other).impl
+
+    def __ne__(self, other: Any) -> bool:
+        if not isinstance(other, AuxiliaryEffectSlot):
+            return NotImplemented
+        return self.impl != (<AuxiliaryEffectSlot> other).impl
+
+    def __gt__(self, other: Any) -> bool:
+        if not isinstance(other, AuxiliaryEffectSlot):
+            return NotImplemented
+        return self.impl > (<AuxiliaryEffectSlot> other).impl
+
+    def __ge__(self, other: Any) -> bool:
+        if not isinstance(other, AuxiliaryEffectSlot):
+            return NotImplemented
+        return self.impl >= (<AuxiliaryEffectSlot> other).impl
+
+    def __bool__(self) -> bool:
+        return <boolean> self.impl
+
+    def set_gain(self, gain: float) -> None:
+        """If set to true, the reverb effect will automatically
+	apply adjustments to the source's send slot gains based
+	on the effect properties.
+
+        Has no effect when using non-reverb effects. Default is true.
+        """
+        return self.impl.set_gain()
+
+    def send_auto(self, sendauto: bool) -> None:
+        """Update the effect slot with a new effect.
+	The given effect object may be altered or
+	destroyed without affecting the effect slot.
+        """
+        return self.impl.set_send_auto()
+
+    # TODO: apply effect
+
+    def destroy(self) -> None:
+        """Destroys the effect slot, returning it to the system.
+        If the effect slot is currently set on a source send,
+        it will be removed first
+        """
+        return self.impl.destroy()
+
+    @property
+    def source_sends(self) -> List[SourceSend]:
```

Comment:
Please define `SourceSend` along with other structs.

Reply:
I failed to write `SourceSend` and apparently `Source` can't be written because it wasn't defined, but since I'm not sure I'll rewrite it before making an issue.